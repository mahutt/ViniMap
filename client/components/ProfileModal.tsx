import React, { useState } from 'react';
import {
  Modal,
  View,
  Text,
  TouchableOpacity,
  StyleSheet,
  Image,
  TextInput,
  FlatList,
} from 'react-native';
import { Feather } from '@expo/vector-icons';

interface ProfileModalProps {
  visible: boolean;
  onClose: () => void;
  userProfile: {
    photoUrl: string;
    name: string;
    email: string;
    calendars: string[];
  };
  onSave: (calendarId: string) => void;
  onSignOut: () => void;
}

const ProfileModal: React.FC<ProfileModalProps> = ({
  visible,
  onClose,
  userProfile,
  onSave,
  onSignOut,
}) => {
  const [calendarId, setCalendarId] = useState('');

  const handleSave = () => {
    onSave(calendarId);
    setCalendarId('');
    onClose();
  };

  return (
    <Modal animationType="slide" transparent={true} visible={visible} onRequestClose={onClose}>
      <View style={styles.modalOverlay}>
        <View style={styles.modalContainer}>
          <View style={styles.modalHeader}>
            <Text style={styles.modalTitle}>Your Profile</Text>
            <TouchableOpacity testID="close-button" onPress={onClose}>
              <Feather name="x" size={24} color="#333" />
            </TouchableOpacity>
          </View>

          <View style={styles.profileSection}>
            <Image source={{ uri: userProfile.photoUrl }} style={styles.profileImage} />
            <Text style={styles.profileName}>{userProfile.name}</Text>
            <Text style={styles.profileEmail}>{userProfile.email}</Text>
          </View>

          <View style={styles.calendarSection}>
            <Text style={styles.sectionTitle}>Your Calendars</Text>
            <FlatList
              data={userProfile.calendars}
              keyExtractor={(item, index) => index.toString()}
              renderItem={({ item }) => (
                <TouchableOpacity style={styles.calendarItem} onPress={() => onSave(item)}>
                  <Feather name="calendar" size={16} color="#852C3A" />
                  <Text style={styles.calendarItemText}>{item}</Text>
                </TouchableOpacity>
              )}
              style={styles.calendarList}
            />
          </View>

          <View style={styles.divider} />

          <View style={styles.inputSection}>
            <Text style={styles.sectionTitle}>Add Calendar ID</Text>
            <TextInput
              style={styles.input}
              placeholder="Enter calendar ID"
              value={calendarId}
              onChangeText={setCalendarId}
            />
            <TouchableOpacity
              style={styles.saveButton}
              onPress={handleSave}
              disabled={calendarId.trim() === ''}>
              <Text style={styles.saveButtonText}>Add Calendar</Text>
            </TouchableOpacity>
          </View>

          <TouchableOpacity style={styles.signOutButton} onPress={onSignOut}>
            <Feather name="log-out" size={16} color="#fff" />
            <Text style={styles.signOutButtonText}>Sign Out</Text>
          </TouchableOpacity>
        </View>
      </View>
    </Modal>
  );
};

const styles = StyleSheet.create({
  modalOverlay: {
    flex: 1,
    backgroundColor: 'rgba(0,0,0,0.5)',
    justifyContent: 'center',
    alignItems: 'center',
  },
  modalContainer: {
    width: '85%',
    backgroundColor: 'white',
    borderRadius: 12,
    padding: 20,
    shadowColor: '#000',
    shadowOffset: {
      width: 0,
      height: 2,
    },
    shadowOpacity: 0.25,
    shadowRadius: 3.84,
    elevation: 5,
  },
  modalHeader: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
    marginBottom: 20,
  },
  modalTitle: {
    fontSize: 20,
    fontWeight: '700',
  },
  profileSection: {
    alignItems: 'center',
    marginBottom: 20,
  },
  profileImage: {
    width: 80,
    height: 80,
    borderRadius: 40,
    marginBottom: 10,
  },
  profileName: {
    fontSize: 18,
    fontWeight: '600',
    marginBottom: 5,
  },
  profileEmail: {
    fontSize: 14,
    color: '#666',
  },
  calendarSection: {
    marginBottom: 20,
  },
  sectionTitle: {
    fontSize: 16,
    fontWeight: '600',
    marginBottom: 10,
  },
  calendarList: {
    maxHeight: 120,
  },
  calendarItem: {
    flexDirection: 'row',
    alignItems: 'center',
    padding: 10,
    backgroundColor: '#f5f5f5',
    borderRadius: 8,
    marginBottom: 8,
  },
  calendarItemText: {
    marginLeft: 10,
    fontSize: 14,
  },
  divider: {
    height: 1,
    backgroundColor: '#e0e0e0',
    marginVertical: 15,
  },
  inputSection: {
    marginBottom: 20,
  },
  input: {
    borderWidth: 1,
    borderColor: '#ddd',
    borderRadius: 8,
    padding: 12,
    marginBottom: 15,
  },
  saveButton: {
    backgroundColor: '#852C3A',
    padding: 12,
    borderRadius: 8,
    alignItems: 'center',
  },
  saveButtonText: {
    color: 'white',
    fontWeight: '600',
  },
  signOutButton: {
    backgroundColor: '#333',
    padding: 12,
    borderRadius: 8,
    alignItems: 'center',
    flexDirection: 'row',
    justifyContent: 'center',
  },
  signOutButtonText: {
    color: 'white',
    fontWeight: '600',
    marginLeft: 8,
  },
});

export default ProfileModal;
