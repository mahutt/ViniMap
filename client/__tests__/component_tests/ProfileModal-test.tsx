import React from 'react';
import { render, fireEvent } from '@testing-library/react-native';
import ProfileModal from '@/components/ProfileModal';

jest.mock('@expo/vector-icons', () => ({
  Feather: 'Feather',
}));

describe('ProfileModal', () => {
  const mockOnClose = jest.fn();
  const mockOnSave = jest.fn();
  const mockOnSignOut = jest.fn();

  const mockUserProfile = {
    photoUrl: 'https://example.com/photo.jpg',
    name: 'John Doe',
    email: 'john.doe@example.com',
    calendars: ['Calendar 1', 'Calendar 2', 'Work Calendar'],
  };

  const defaultProps = {
    visible: true,
    onClose: mockOnClose,
    userProfile: mockUserProfile,
    onSave: mockOnSave,
    onSignOut: mockOnSignOut,
  };

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('matches the snapshot', () => {
    const { toJSON } = render(<ProfileModal {...defaultProps} />);
    expect(toJSON()).toMatchSnapshot();
  });

  it('renders correctly with user profile data', () => {
    const { getByText, getByPlaceholderText } = render(<ProfileModal {...defaultProps} />);

    expect(getByText('John Doe')).toBeTruthy();
    expect(getByText('john.doe@example.com')).toBeTruthy();

    expect(getByText('Calendar 1')).toBeTruthy();
    expect(getByText('Calendar 2')).toBeTruthy();
    expect(getByText('Work Calendar')).toBeTruthy();

    expect(getByPlaceholderText('Enter calendar ID')).toBeTruthy();
    expect(getByText('Add Calendar')).toBeTruthy();
    expect(getByText('Sign Out')).toBeTruthy();
  });

  it('handles calendar selection correctly', () => {
    const { getByText } = render(<ProfileModal {...defaultProps} />);

    const calendarItem = getByText('Work Calendar');
    fireEvent.press(calendarItem);

    expect(mockOnSave).toHaveBeenCalledWith('Work Calendar');
  });

  it('handles adding a new calendar correctly', () => {
    const { getByText, getByPlaceholderText } = render(<ProfileModal {...defaultProps} />);

    const input = getByPlaceholderText('Enter calendar ID');
    const addButton = getByText('Add Calendar');

    fireEvent.changeText(input, 'New Calendar');
    fireEvent.press(addButton);

    expect(mockOnSave).toHaveBeenCalledWith('New Calendar');
    expect(mockOnClose).toHaveBeenCalledTimes(1);
  });

  it('handles sign out correctly', () => {
    const { getByText } = render(<ProfileModal {...defaultProps} />);

    const signOutButton = getByText('Sign Out');
    fireEvent.press(signOutButton);

    expect(mockOnSignOut).toHaveBeenCalledTimes(1);
  });

  it('handles close button correctly', () => {
    const { getByTestId } = render(<ProfileModal {...defaultProps} />);

    // You might need to add a testID to your close button in the component
    // For example: <TouchableOpacity testID="close-button" onPress={onClose}>
    const closeButton = getByTestId('close-button');
    fireEvent.press(closeButton);

    expect(mockOnClose).toHaveBeenCalledTimes(1);
  });

  it('does not render when visible prop is false', () => {
    const { queryByText } = render(<ProfileModal {...defaultProps} visible={false} />);

    expect(queryByText('Your Profile')).toBeNull();
  });
});
