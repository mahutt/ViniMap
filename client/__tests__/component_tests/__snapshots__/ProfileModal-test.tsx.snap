// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProfileModal matches the snapshot 1`] = `
<Modal
  animationType="slide"
  hardwareAccelerated={false}
  onRequestClose={[MockFunction]}
  transparent={true}
  visible={true}
>
  <View
    style={
      {
        "alignItems": "center",
        "backgroundColor": "rgba(0,0,0,0.5)",
        "flex": 1,
        "justifyContent": "center",
      }
    }
  >
    <View
      style={
        {
          "backgroundColor": "white",
          "borderRadius": 12,
          "elevation": 5,
          "padding": 20,
          "shadowColor": "#000",
          "shadowOffset": {
            "height": 2,
            "width": 0,
          },
          "shadowOpacity": 0.25,
          "shadowRadius": 3.84,
          "width": "85%",
        }
      }
    >
      <View
        style={
          {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "space-between",
            "marginBottom": 20,
          }
        }
      >
        <Text
          style={
            {
              "fontSize": 20,
              "fontWeight": "700",
            }
          }
        >
          Your Profile
        </Text>
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "opacity": 1,
            }
          }
          testID="close-button"
        >
          <Feather
            color="#333"
            name="x"
            size={24}
          />
        </View>
      </View>
      <View
        style={
          {
            "alignItems": "center",
            "marginBottom": 20,
          }
        }
      >
        <Image
          source={
            {
              "uri": "https://example.com/photo.jpg",
            }
          }
          style={
            {
              "borderRadius": 40,
              "height": 80,
              "marginBottom": 10,
              "width": 80,
            }
          }
        />
        <Text
          style={
            {
              "fontSize": 18,
              "fontWeight": "600",
              "marginBottom": 5,
            }
          }
        >
          John Doe
        </Text>
        <Text
          style={
            {
              "color": "#666",
              "fontSize": 14,
            }
          }
        >
          john.doe@example.com
        </Text>
      </View>
      <View
        style={
          {
            "marginBottom": 20,
          }
        }
      >
        <Text
          style={
            {
              "fontSize": 16,
              "fontWeight": "600",
              "marginBottom": 10,
            }
          }
        >
          Your Calendars
        </Text>
        <RCTScrollView
          data={
            [
              "Calendar 1",
              "Calendar 2",
              "Work Calendar",
            ]
          }
          getItem={[Function]}
          getItemCount={[Function]}
          keyExtractor={[Function]}
          onContentSizeChange={[Function]}
          onLayout={[Function]}
          onMomentumScrollBegin={[Function]}
          onMomentumScrollEnd={[Function]}
          onScroll={[Function]}
          onScrollBeginDrag={[Function]}
          onScrollEndDrag={[Function]}
          removeClippedSubviews={false}
          renderItem={[Function]}
          scrollEventThrottle={0.0001}
          stickyHeaderIndices={[]}
          style={
            {
              "maxHeight": 120,
            }
          }
          viewabilityConfigCallbackPairs={[]}
        >
          <View>
            <View
              onFocusCapture={[Function]}
              onLayout={[Function]}
              style={null}
            >
              <View
                accessibilityState={
                  {
                    "busy": undefined,
                    "checked": undefined,
                    "disabled": undefined,
                    "expanded": undefined,
                    "selected": undefined,
                  }
                }
                accessibilityValue={
                  {
                    "max": undefined,
                    "min": undefined,
                    "now": undefined,
                    "text": undefined,
                  }
                }
                accessible={true}
                collapsable={false}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  {
                    "alignItems": "center",
                    "backgroundColor": "#f5f5f5",
                    "borderRadius": 8,
                    "flexDirection": "row",
                    "marginBottom": 8,
                    "opacity": 1,
                    "padding": 10,
                  }
                }
              >
                <Feather
                  color="#852C3A"
                  name="calendar"
                  size={16}
                />
                <Text
                  style={
                    {
                      "fontSize": 14,
                      "marginLeft": 10,
                    }
                  }
                >
                  Calendar 1
                </Text>
              </View>
            </View>
            <View
              onFocusCapture={[Function]}
              onLayout={[Function]}
              style={null}
            >
              <View
                accessibilityState={
                  {
                    "busy": undefined,
                    "checked": undefined,
                    "disabled": undefined,
                    "expanded": undefined,
                    "selected": undefined,
                  }
                }
                accessibilityValue={
                  {
                    "max": undefined,
                    "min": undefined,
                    "now": undefined,
                    "text": undefined,
                  }
                }
                accessible={true}
                collapsable={false}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  {
                    "alignItems": "center",
                    "backgroundColor": "#f5f5f5",
                    "borderRadius": 8,
                    "flexDirection": "row",
                    "marginBottom": 8,
                    "opacity": 1,
                    "padding": 10,
                  }
                }
              >
                <Feather
                  color="#852C3A"
                  name="calendar"
                  size={16}
                />
                <Text
                  style={
                    {
                      "fontSize": 14,
                      "marginLeft": 10,
                    }
                  }
                >
                  Calendar 2
                </Text>
              </View>
            </View>
            <View
              onFocusCapture={[Function]}
              onLayout={[Function]}
              style={null}
            >
              <View
                accessibilityState={
                  {
                    "busy": undefined,
                    "checked": undefined,
                    "disabled": undefined,
                    "expanded": undefined,
                    "selected": undefined,
                  }
                }
                accessibilityValue={
                  {
                    "max": undefined,
                    "min": undefined,
                    "now": undefined,
                    "text": undefined,
                  }
                }
                accessible={true}
                collapsable={false}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  {
                    "alignItems": "center",
                    "backgroundColor": "#f5f5f5",
                    "borderRadius": 8,
                    "flexDirection": "row",
                    "marginBottom": 8,
                    "opacity": 1,
                    "padding": 10,
                  }
                }
              >
                <Feather
                  color="#852C3A"
                  name="calendar"
                  size={16}
                />
                <Text
                  style={
                    {
                      "fontSize": 14,
                      "marginLeft": 10,
                    }
                  }
                >
                  Work Calendar
                </Text>
              </View>
            </View>
          </View>
        </RCTScrollView>
      </View>
      <View
        style={
          {
            "backgroundColor": "#e0e0e0",
            "height": 1,
            "marginVertical": 15,
          }
        }
      />
      <View
        style={
          {
            "marginBottom": 20,
          }
        }
      >
        <Text
          style={
            {
              "fontSize": 16,
              "fontWeight": "600",
              "marginBottom": 10,
            }
          }
        >
          Add Calendar ID
        </Text>
        <TextInput
          onChangeText={[Function]}
          placeholder="Enter calendar ID"
          style={
            {
              "borderColor": "#ddd",
              "borderRadius": 8,
              "borderWidth": 1,
              "marginBottom": 15,
              "padding": 12,
            }
          }
          value=""
        />
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": true,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={false}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "alignItems": "center",
              "backgroundColor": "#852C3A",
              "borderRadius": 8,
              "opacity": 1,
              "padding": 12,
            }
          }
        >
          <Text
            style={
              {
                "color": "white",
                "fontWeight": "600",
              }
            }
          >
            Add Calendar
          </Text>
        </View>
      </View>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignItems": "center",
            "backgroundColor": "#333",
            "borderRadius": 8,
            "flexDirection": "row",
            "justifyContent": "center",
            "opacity": 1,
            "padding": 12,
          }
        }
      >
        <Feather
          color="#fff"
          name="log-out"
          size={16}
        />
        <Text
          style={
            {
              "color": "white",
              "fontWeight": "600",
              "marginLeft": 8,
            }
          }
        >
          Sign Out
        </Text>
      </View>
    </View>
  </View>
</Modal>
`;
